File: 62-unique-paths/62-unique-paths.cpp
class Solution {
public:
    vector<vector<int>>dp;
    int solve(int n, int m){
        if(n==1 or m==1){
            return 1;
        }
        if(dp[n][m]!=-1){
            return dp[n][m];
        }
        return dp[n][m]=solve(n-1,m)+solve(n,m-1);
    }
    int uniquePaths(int n, int m) {
        dp.resize(n+1,vector<int>(m+1,-1));
        return solve(n,m);
    }
};

File: 62-unique-paths/62-unique-paths.java
class Solution {
    public int uniquePaths(int n, int m) {
        Integer[][] dp = new Integer[n+1][m+1];
        for(int i=1;i<=n;i++){
            dp[i][1]=1;
        }
        for(int j=1;j<=m;j++){
            dp[1][j]=1;
        }
        for(int i=2;i<=n;i++){
            for(int j=2;j<=m;j++){
                dp[i][j]=dp[i-1][j]+dp[i][j-1];
            }
        }
        return dp[n][m];
    }
}